server/map_rpc_server.go

```diff
@@ -128,45 +135,50 @@ func (t *TrillianMapServer) getLeavesByRevision(ctx context.Context, mapID int64
 	if err := mapRoot.UnmarshalBinary(root.MapRoot); err != nil {
 		return nil, err
 	}
+	revision = int64(mapRoot.Revision)
 
-	smtReader := merkle.NewSparseMerkleTreeReader(int64(mapRoot.Revision), hasher, tx)
+	leaves, err := tx.Get(ctx, revision, indices)
+	if err != nil {
+		return nil, fmt.Errorf("could not fetch leaves: %v", err)
+	}
+	leavesByIndex := make(map[string]*trillian.MapLeaf)
+	for i, l := range leaves {
+		leavesByIndex[string(l.Index)] = &leaves[i]
+	}
+	glog.V(1).Infof("%v: wanted %v leaves, found %v", mapID, len(indices), len(leaves))
 
-	inclusions := make([]*trillian.MapLeafInclusion, 0, len(indices))
-	found := 0
+	// Add empty leaf values for indexes that were not returned.
 	for _, index := range indices {
-		if err := checkIndexSize(index, hasher); err != nil {
-			return nil, err
-		}
-		// Fetch the leaf if it exists.
-		leaves, err := tx.Get(ctx, int64(mapRoot.Revision), [][]byte{index})
-		if err != nil {
-			return nil, fmt.Errorf("could not fetch leaf %x: %v", index, err)
-		}
-		var leaf *trillian.MapLeaf
-		if len(leaves) == 1 {
-			leaf = &leaves[0]
-			found++
-		} else {
-			// Empty leaf for proof of non-existence.
-			leaf = &trillian.MapLeaf{
-				Index:     index,
-				LeafValue: nil,
-				LeafHash:  nil, // equivalent to hasher.HashEmpty(mapID, index, 0)
-			}
-		}
-
-		// Fetch the proof regardless of whether the leaf exists.
-		proof, err := smtReader.InclusionProof(ctx, int64(mapRoot.Revision), index)
-		if err != nil {
-			return nil, fmt.Errorf("could not get inclusion proof for leaf %x: %v", index, err)
+		if _, ok := leavesByIndex[string(index)]; !ok {
+			leavesByIndex[string(index)] = &trillian.MapLeaf{Index: index}
 		}
+	}
 
-		inclusions = append(inclusions, &trillian.MapLeafInclusion{
-			Leaf:      leaf,
-			Inclusion: proof,
-		})
+	// Fetch inclusion proofs in parallel.
+	smtReader := merkle.NewSparseMerkleTreeReader(int64(mapRoot.Revision), hasher, tx)
+	inclusions := make([]*trillian.MapLeafInclusion, len(indices))
+	errs := make(chan error, len(indices))
+	var wg sync.WaitGroup
+	wg.Add(len(indices))
+	for i, index := range indices {
+		go func(i int, index []byte) {
```

Comment:
I am concerned about 1 goroutine per index approach w.r.t. scale. Would introducing a batch request instead be more reliable?

Reply:
A batch index request would be infinitely better, but I'm working with what I've got.

Reply:
Could you please leave a TODO to improve this?

Reply:
Maybe even leave the issue open, to address it in a bit.